src/server.rs

```diff
@@ -1,106 +1,96 @@
 use opentelemetry::trace::{Span, TraceContextExt, Tracer};
 use opentelemetry::{Context, KeyValue};
 use serde::{Deserialize, Serialize};
 
-use tokio::sync::Mutex;
 use tracing::{debug, error, info};
 
 use jsonrpsee::proc_macros::rpc;
 
-const CACHE_SIZE: usize = 100;
+const CACHE_SIZE: u64 = 100;
 
 pub struct PayloadTraceContext {
-    tracer: Arc<BoxedTracer>,
-    block_hash_to_payload_ids: Arc<Mutex<LruCache<B256, Vec<PayloadId>>>>,
-    payload_id_to_span: Arc<Mutex<LruCache<PayloadId, Arc<BoxedSpan>>>>,
-    local_to_external_payload_ids: Arc<Mutex<LruCache<PayloadId, PayloadId>>>,
+    tracer: BoxedTracer,
+    block_hash_to_payload_ids: Cache<B256, Vec<PayloadId>>,
+    payload_id_to_span: Cache<PayloadId, Arc<BoxedSpan>>,
+    local_to_external_payload_ids: Cache<PayloadId, PayloadId>,
 }
 
 impl PayloadTraceContext {
     fn new() -> Self {
         PayloadTraceContext {
-            tracer: Arc::new(global::tracer("rollup-boost")),
-            block_hash_to_payload_ids: Arc::new(Mutex::new(LruCache::new(
-                NonZero::new(CACHE_SIZE).unwrap(),
-            ))),
-            payload_id_to_span: Arc::new(Mutex::new(LruCache::new(
-                NonZero::new(CACHE_SIZE).unwrap(),
-            ))),
-            local_to_external_payload_ids: Arc::new(Mutex::new(LruCache::new(
-                NonZero::new(CACHE_SIZE).unwrap(),
-            ))),
+            tracer: global::tracer("rollup-boost"),
+            block_hash_to_payload_ids: Cache::new(CACHE_SIZE),
+            payload_id_to_span: Cache::new(CACHE_SIZE),
+            local_to_external_payload_ids: Cache::new(CACHE_SIZE),
         }
     }
 
-    async fn store(&self, payload_id: PayloadId, parent_hash: B256, parent_span: BoxedSpan) {
-        let mut store = self.payload_id_to_span.lock().await;
-        store.put(payload_id, Arc::new(parent_span));
-        let mut store = self.block_hash_to_payload_ids.lock().await;
-        if let Some(payload_ids) = store.get_mut(&parent_hash) {
-            payload_ids.push(payload_id);
-        } else {
-            store.put(parent_hash, vec![payload_id]);
-        }
+    fn store(&self, payload_id: PayloadId, parent_hash: B256, parent_span: BoxedSpan) {
+        self.payload_id_to_span
+            .insert(payload_id, Arc::new(parent_span));
+        self.block_hash_to_payload_ids
+            .entry(parent_hash)
+            .and_upsert_with(|o| match o {
+                Some(e) => {
+                    let mut payloads = e.into_value();
+                    payloads.push(payload_id);
+                    payloads
+                }
+                None => {
+                    vec![payload_id]
+                }
+            });
     }
 
-    async fn retrieve_by_parent_hash(&self, parent_hash: &B256) -> Option<Vec<Arc<BoxedSpan>>> {
-        let mut block_hash_to_payload_ids = self.block_hash_to_payload_ids.lock().await;
-        let mut payload_id_to_span = self.payload_id_to_span.lock().await;
```

Comment:
Deadlock can occur here, as locks are acquired in opposite order as in `fn store` above.